pkg/deploy/controller/deploymentconfig/controller.go

```diff
@@ -263,17 +264,43 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments *k
 		}
 	}
 
-	return c.updateStatus(config)
+	return c.updateStatus(config, existingDeployments.Items)
 }
 
-func (c *DeploymentConfigController) updateStatus(config *deployapi.DeploymentConfig) error {
-	if config.Generation > config.Status.ObservedGeneration {
-		config.Status.ObservedGeneration = config.Generation
-	}
-	if _, err := c.osClient.DeploymentConfigs(config.Namespace).UpdateStatus(config); err != nil {
+func (c *DeploymentConfigController) updateStatus(config *deployapi.DeploymentConfig, deployments []kapi.ReplicationController) error {
+	newStatus, err := calculateStatus(*config, deployments)
+	if err != nil {
 		glog.V(2).Infof("Cannot update the status for %q: %v", deployutil.LabelForDeploymentConfig(config), err)
 		return transientError(fmt.Sprintf("cannot update the status for %q - requeuing", deployutil.LabelForDeploymentConfig(config)))
 	}
+
+	if !reflect.DeepEqual(*newStatus, config.Status) {
+		config.Status = newStatus
+		if _, err := c.osClient.DeploymentConfigs(config.Namespace).UpdateStatus(config); err != nil {
+			glog.V(2).Infof("Cannot update the status for %q: %v", deployutil.LabelForDeploymentConfig(config), err)
+			return transientError(fmt.Sprintf("cannot update the status for %q - requeuing", deployutil.LabelForDeploymentConfig(config)))
+		}
+		glog.V(4).Infof("Updated the status for %q (observed generation: %d)", deployutil.LabelForDeploymentConfig(config), config.Status.ObservedGeneration)
+	}
 	glog.V(4).Infof("Updated the status for %q (observed generation: %d)", deployutil.LabelForDeploymentConfig(config), config.Status.ObservedGeneration)
 	return nil
 }
+
+func calculateStatus(config deployapi.DeploymentConfig, deployments []kapi.ReplicationController) (*deployapi.DeploymentConfigStatus, error) {
```

Comment:
The actual config is only used for the error messages, so I think this method should really just return the three counts and not consume the config at all, or update the given config (and consume it as a pointer).